[ws-manager] Re-create workspace pods on rejection

geropl opened the pull request:
## Description
This PR enables `ws-manager` to **re-create** a workspace pod when it detects that the pod is stuck. This i
s done by stopping the failed workspace, wait a configurable timeout (15s default), and create a pod with the same name/ID again.

The root cause is an unfixed race condition in kubelet: node labels are out of sync, and suddenly the Pod - although already scheduled to a node - does no longer fit onto the node, so the kubelet marks it as "rejected": A situation, in which it is "stuck".
It shows up in both AWS and GCP, and has a ~1% chance of occurring (from our rough investigation).

This change makes it so that ws-manager detects this condition (which is luckily very straight forward and clearly distinguishable), and once detected, marks it with the `PodRejected` condition. From there on, workspace is treated differently. This ensures that this change remains (nearly) zero-impact for the other 99% of workspaces (with the exception of some changes to context handling in ws-daemon's pod control loop, see below).

To make sure this change does not break anything it has been thoroughly stress-tested using a `loadgen` + a `breaker` program which "rejects" every workspace that is created exactly once (see the "DEV" commit).

### TODO
 - [x] add happy-path test
 - [x] manual tests
   - [x] regular workspace starts
   - [x] prebuild workspace starts
   - [x] regular workspace + forced "pod rejection"
 - [x] add "WorkspaceStatus" test
 - [x] do load test, as we saw the CPU variant on GCP as well ([docs](https://www.notion.so/gitpod/How-to-create-delete-ephemeral-workspace-cluster-2e636dae17d148a5a7de5afd54566479) and [docs](https://www.notion.so/gitpod/Run-loadgen-against-a-workspace-preview-055191214e5f453cb8a961a7128e9b62))
 - [x] discuss with folks to make sure we don't miss any edges cases 
 - [x] remove debug logging (UPPER CASE)
 - [x] :exclamation: drop DEV commit before merging

## Discussion: "Cool, but it's a big change. Is this worth it?"

Yes, I think so. And that's because:
 - a) it's fairly isolated (the change does impose very little risk to the majority/99% of workspaces) which continue to function :100:% as they do today.
 - b) it fixes a long-standing bug (it's a kubernetes bug, granted, but it directly affects Gitpod usability)

## Changes in Detail

The general theme is that we maintain all the existing logic of starting and stopping workspaces. The only conceptual changes are:
- Stopping a workspace with `PodRejected` works a bit differently, as a) we don't care about a backup, b) it might fail at _any_ point during `Pending`/`Creating` and c) we need to cleanup _everything_ we know about that workspace in ws-daemon, as there might be a new pod with the exact same ID in the near future, soon.
  - the brunt of the changes are actually related to enabling `ws-daemon` to "wipe" a workspace, in a synchronized way ( this is _only_ exercised in this failure condition!)
  - this includes using a dedicated condition to communicated this being done (next to `BackupFailure`/`BackupComplete`): `StateWiped`
 - once a workspace with `PodRejected` has been stopped, it gets eventually re-created

### ws-manager

Two new config fields:
 - `PodRecreationMaxRetries`: (default: `3`): how often we are trying to re-create a given Workspace
 - `PodRecreationBackoff` (default: `15s`): how long two wait before creating a new pod after the old one has been deleted

The workspace CRD gets two new fields:
 - `.status.podRecreated`: the number of times a workspace has gone through the re-creation logic (usually `0`)
 - `.status.podDeletionTime`: The time the workspace pod was seen removed the first time by ws-manager (set in `status.go`)
   - this is used to make sure 
There is a 3rd field `podStarts` which has been there since forever, which is finally used now. :slightly_smiling_face: 

Two new conditions got introduced to, which are :
 - `PodRejected` (set in `status.go`)
   - `true`: when the pod is `failing` (pending -> stopped)
   - `false`: when the pod is re-created
 - `StateWiped` (set in ws-daemon, and checked in `status.go`)
   - used to synchronize the "workspace state wiped" state from ws-daemon back to ws-manager. Only once this is set a "rejected workspace" is considered "stopped".

### ws-daemon

**Workspace control loop**

 On `handleWorkspaceStop`, we call `doWipeWorkspace` instead of `doWorkspaceContentBackup`. It calls `workspace_operations.WipeWorkspace`, and sets the `StateWiped` condition accordingly.

`WipeWorkspace` does:
 - calls the regular "dispose" hooks
 - calls an additional "dispose" hook, which tries to call the IWS, and if successful, runs `WipingTeardown`
   - this unmount the shiftfs mounts from inside the workspace
   - :information_source: this might be a change we actually can use for all workspaces. But that's a separate discussion.
 - calls `dispatch.DisposeWorkspace` (see below)
 - disposes all stored metadata (like on regular shutdown)

**Pod control loop**

This control loop is fueled by a `PodInformer`, and before this change it's shutdown mechanics were not sync'ed with the `Workspace` control loop, and timing worked on "best-effort" basis.

 `dispatch.DisposeWorkspace` got added (which again is only called in the `PodRejected` case, cmp. above), which:
 - cancels the context of _all_ running go routines
 - waits for all of them to report being cancelled (`WaitGroup`)
 - calls `containerRuntime.DisposeContainer`, which clears some internally held ID mappings between Workspace, Pod and Container
 - :information_source: blocks future pod updates (excluding "deletions") from being processed (pod is being identified by `.Name + .CreationTimestamp`!)

## Diagram

I tried to capture the new state changes in a diagram ([source](https://excalidraw.com/#json=KBVIH09nYphnvCsTMMPx4,2_YiXbRidQC76DejiUbhag)):
![Untitled-2024-10-08-1358](https://github.com/user-attachments/assets/eed9d58e-2b53-4309-ac05-9329dc04ace2)


## Related Issue(s)
<!-- List the issue(s) this PR solves -->
Fixes ENT-530

## How to test

### Check tests are sensible and :green_circle: 

### Manual: regular workspace

### Manual: prebuild workspace

### Manual: using `loadgen` and the "rejection"-program

Loom of my testing is here: https://www.loom.com/share/e2eae31965dc48829b415c81f3005f1b?sid=6b4acf8d-06d0-4d21-b3e2-ef3182f8e974
Summary screen: 
![image](https://github.com/user-attachments/assets/532600ca-b517-462a-a1d0-da9dc68ecc95)


1. open a workspace on this branch,
2. connect to the `ephemeral-gpl4` cluster (announce here so you don't block others)
3. prepare to run a loadgen test as usual
4. in parallel, start "rejection" tool: `cd scheduler && go run main.go -kubeconfig ~/.kube/config`
5. run `loadgen`
6. notice how all workspaces come up nicely :relieved: 

TODO(gpl): add loom so others don't have to do this

## Documentation
<!--
Does this PR require updates to the documentation at www.gitpod.io/docs?
* Yes
  * 1. Please create a docs issue: https://github.com/gitpod-io/website/issues/new?labels=documentation&template=DOCS-NEW-FEATURE.yml&title=%5BDocs+-+New+Feature%5D%3A+%3Cyour+feature+name+here%3E
  * 2. Paste the link to the docs issue below this comment
* No
  * Are you sure? If so, nothing to do here.
-->

#### Preview status

<p>Gitpod was successfully deployed to your preview environment.</p>
<ul>
	<li><b>🏷️ Name</b> - gpl-530-retry</li>
	<li><b>🔗 URL</b> - <a href="https://gpl-530-retry.preview.gitpod-dev.com/workspaces" target="_blank">gpl-530-retry.preview.gitpod-dev.com/workspaces</a>.</li>
	<li><b>📚 Documentation</b> - See our <a href="https://www.notion.so/gitpod/6debd359591b43688b52f76329d04010#7c1ce80ab31a41e29eff2735e38eec39" target="_blank">internal documentation</a> for information on how to interact with your preview environment.</li>
	<li><b>📦 Version</b> - gpl-530-retry-gha.30031</li>
	<li><b>🗒️ Logs</b> - <a href="https://console.cloud.google.com/logs/query;query=jsonPayload.kubernetes.host%3D%22preview-gpl-530-retry%22%0A%0A--%20Filter%20on%20service:%0A--%20jsonPayload.serviceContext.service%3D%22ws-manager-mk2%22%0A;duration=P1D?project=gitpod-dev-preview" target="_blank">GCP Logs Explorer</a></li>
</ul>

## Build Options

<details>
<summary>Build</summary>

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`
</details>

<details>
<summary>Publish</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer</summary>

- [ ] analytics=segment
- [ ] with-dedicated-emulation
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

<details>
<summary>Preview Environment / Integration Tests</summary>

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [ ] /werft with-large-vm
- [x] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [x] /werft preemptible
      Saves cost. Untick this only if you're really sure you need a non-preemtible machine.
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`. If enabled, `with-preview` and `with-large-vm` will be enabled.
- [ ] with-monitoring
</details>

/hold

geropl commented:
Still work in progress

WVerlaek commented on components/ws-manager-mk2/controllers/workspace_controller.go:
```diff
@@ -221,6 +219,55 @@ func (r *WorkspaceReconciler) actOnStatus(ctx context.Context, workspace *worksp
+			workspace.Status.PodRecreated = sc.PodRecreated + 1
+			workspace.Status.SetCondition(workspacev1.NewWorkspaceConditionPodRejected(fmt.Sprintf("Recreating pod... (%d retry)", workspace.Status.PodRecreated), metav1.ConditionFalse))
+
+			if err := r.Status().Patch(ctx, workspace, patch); err != nil {
```
I'm not super convinced this should be a `Patch`, an `Update` would catch acting on an out-of-date status by returning a conflict error

geropl replied:
Hm, with how the current code works, I think I agree. Will try it out. :+1:

geropl replied:
`Update` worked flawlessly, will keep it!

WVerlaek reviewed (commented):
overall nice work! 
Left a few comments/questions, didn't look too deeply into the ws-daemon changes

geropl commented on components/ws-daemon/pkg/daemon/markunmount.go:
```diff
@@ -101,8 +101,24 @@ func (c *MarkUnmountFallback) WorkspaceUpdated(ctx context.Context, ws *dispatch
+			c.mu.Unlock()
+		}()
+
+		wait := time.NewTicker(ttl)
+		defer wait.Stop()
+		select {
+		case <-ctx.Done():
+			return
+		case <-wait.C:
```
> <-time.After(ttl)

Doesn't this leak a timer in case it does not fire? It's the vast minority of cases, granted, but the reason I preferred the explicit way (`NewTicker` + `defer ...Stop()`)

WVerlaek replied:
ah only since Go 1.23 it doesn't leak, and daemon is on 1.22:

```
// As of Go 1.23, the garbage collector can recover unreferenced,
// unstopped timers. There is no reason to prefer NewTimer when After will do.
```

somewhere it was missing a defer though

geropl replied:
> ah only since Go 1.23 it doesn't leak

:bulb: TIL!

> somewhere it was missing a defer though

Will have a look :eyes:

geropl replied:
Found and fixed.

geropl commented on components/ws-daemon/pkg/controller/workspace_operations.go:
```diff
@@ -285,6 +293,53 @@ func (wso *DefaultWorkspaceOperations) DeleteWorkspace(ctx context.Context, inst
+		}
+	}()
+
+	timeoutT := time.NewTicker(10 * time.Second)
+	select {
+	case <-timeoutT.C:
+	case <-removedChan:
+		log.Debug("successfully removed workspace daemon directory")
+	}
```
Same discussion as https://github.com/gitpod-io/gitpod/pull/20243#discussion_r1838081913

geropl commented on components/ws-manager-mk2/controllers/workspace_controller.go:
```diff
@@ -221,6 +219,55 @@ func (r *WorkspaceReconciler) actOnStatus(ctx context.Context, workspace *worksp
+		case workspace.Status.Phase == workspacev1.WorkspacePhaseStopped && workspace.IsConditionTrue(workspacev1.WorkspaceConditionPodRejected):
+			if workspace.Status.PodRecreated > r.Config.PodRecreationMaxRetries {
+				workspace.Status.SetCondition(workspacev1.NewWorkspaceConditionPodRejected(fmt.Sprintf("Pod reached maximum recreations %d, failing", workspace.Status.PodRecreated), metav1.ConditionFalse))
+				return ctrl.Result{Requeue: true}, nil // requeue so we end up in the "Stopped" case below
```
> Should we instead e.g. set the Failed condition on the workspace?

We leave the Failed condition untouched, which effectively leaves be set to the error that caused the last pod to fail, which is a good thing IMO.

> would be nice to have a unit test for this case 🙏

I agree... :spiral_notepad:

geropl replied:
> would be nice to have a unit test for this case 🙏

@WVerlaek Would love to sync with you tomorrow on ideas how to do this. Left a comment [here](https://github.com/gitpod-io/gitpod/commit/002f741c0b83dbdf918ccc57b4ea4f8295350741#r149000229).

geropl commented on scheduler/main.go:
```diff
@@ -0,0 +1,148 @@
+package main
```
I thought about removing this tool, as it's of very limited use.

Do you think we should keep it?

geropl replied:
Adopted it under `dev/rejector`

geropl commented on components/ws-daemon/pkg/dispatch/dispatch.go:
```diff
@@ -99,6 +105,20 @@ func GetFromContext(ctx context.Context) *Dispatch {
+	return ctx.Value(contextDispatchWaitGroup).(*sync.WaitGroup)
+}
+
+func IsCancelled(ctx context.Context) bool {
+	return context.Cause(ctx) != nil
+}
```
There was a reason back then to check `Cause`, but I don't recall rn. :see_no_evil: Will have a 2nd look...

Thank you for the thorough review! :heart_hands:

geropl replied:
Done.

geropl commented on components/ws-daemon/pkg/cgroup/cgroup.go:
```diff
@@ -79,8 +79,11 @@ func (host *PluginHost) WorkspaceAdded(ctx context.Context, ws *dispatch.Workspa
-	cgroupPath, err := disp.Runtime.ContainerCGroupPath(context.Background(), ws.ContainerID)
+	cgroupPath, err := disp.Runtime.ContainerCGroupPath(ctx, ws.ContainerID)
 	if err != nil {
+		if err == context.Canceled {
```
Done.

geropl commented on components/ws-daemon/pkg/netlimit/netlimit.go:
```diff
@@ -141,12 +145,18 @@ func (c *ConnLimiter) limitWorkspace(ctx context.Context, ws *dispatch.Workspace
 		}
 	}, nsinsider.EnterMountNS(false), nsinsider.EnterNetNS(true))
 	if err != nil {
+		if errors.Is(err, context.Canceled) {
```
Ah, that was the reason I initially used a helper function to do `errors.Is(context.Cause(ctx), context.Canceled)` in some places :see_no_evil: 
Re-added it here.

Thank you @Furisto for the thorough review! :bow:

geropl commented on components/ws-daemon/pkg/controller/workspace_operations.go:
```diff
@@ -290,6 +298,51 @@ func (wso *DefaultWorkspaceOperations) DeleteWorkspace(ctx context.Context, inst
+		}
+	}()
+
+	timeout := time.NewTicker(10 * time.Second)
```
Added a comment.

tl;dr: If we don't have the timeout, there is a chance that that this operation blocks forever (at least before I fixed the mark unmounting). By adding the timeout guarantee it never blocks the workspace from being deleted (because that depends on the "wiping" to finish.

We trade it for the slight chance of not wiping cleanly, which _might_ become a problem iff the new workspace pod is scheduled to the same node again. Which I deem ok, because then we are talking about an error case in the 0.1-0.01% range, which is already way better than the original 1%.
All of this, again, only if the mark unmounting did not fix this problem in the first place. :+1:

geropl commented:
@iQQBot I added the loom for the load test (incl. rejector running, so "breaking" every workspace pod once), and addressed all the other comments. Would be great if you could review, and do a manual smoke test on the preview (once it's there again) that the "normal case" still works as expected.

Two things I still need to test (tomorrow):
 - have `dev/rejector` running against the preview
 - manually (+ via prebuild) start workspaces
 - check that the UX does not break :+1: 

Once that's done, we can merge. :slightly_smiling_face:

geropl commented:
Done testing! :partying_face: :tada:

geropl commented on components/ws-manager-mk2/controllers/status.go:
```diff
@@ -123,6 +140,16 @@ func (r *WorkspaceReconciler) updateWorkspaceStatus(ctx context.Context, workspa
 
+	if failure != "" && !workspace.IsConditionTrue(workspacev1.WorkspaceConditionPodRejected) {
+		// Check: A situation where we want to retry?
+		if isPodRejected(pod) {
```
> The kubelet will also attempt to re-admit the pod upon kubelet restart

Ah, did not know, thanks for sharing!

My approach was that today, these cases end as "failed". So whatever I do, I prevent these from happening now, by trying to restart. If what you say is applicable, yes, there might be data loss - but so was before (yes, manual recovery is a thing, but not in Dedicated).

geropl replied:
> If the pod is simply marked as failed, it will continue to perform the standard backup operation.

You are right indeed. :dart: 

We can add that in a follow-up, though.

geropl replied:
Linear issue: https://linear.app/gitpod/issue/CLC-942/[ws-manager]-avoid-podrejection-on-kubelet-restart

geropl commented:
> A question: If content-init has already started and the pod re-schedule at same node, and content is big enough, what will happen? Is there a possibility of a data overlap issue?

The approach in this PR is the "wiping mode" as you found out, where first all running handlers etc. are stopped, and finally all content directories are removed. The oprations are synchronized on multiple levels (one e.g. for content-init), so this _should_* not happen.

*: saying "should", because the ws-daemon code was surprinsingly complex, as we have three layers of feeback loops (k8s workspace, k8s pod, containerd) we have to synchronize. There still might be a whole somewhere, but after the extensive testing, I'm absolutely sure we are taking about 0.01% here, instead of the 1% we have atm.

geropl commented:
/unhold

geropl commented:
> content-init is a special case that executes using another process and has no related context to control (i.e. the context is the context of ws-deamon).

But content init happens in sync. And all workspace-state related changes are handled by `workspace_controller.go`, using the controller framework by k8s, which I understood handles one event at time (e.g. handles synchronization).
